src/main/java/org/primefaces/component/inputnumber/InputNumberRenderer.java

```diff
@@ -67,25 +69,41 @@ public void decode(FacesContext context, UIComponent component) {
         String inputId = inputNumber.getClientId(context) + "_hinput";
         String submittedValue = context.getExternalContext().getRequestParameterMap().get(inputId);
 
-        if (submittedValue != null) {
-
-            BigDecimal value = new BigDecimal(submittedValue);
-            if (!ComponentUtils.isValueBlank(inputNumber.getMinValue())) {
-                BigDecimal min = new BigDecimal(inputNumber.getMinValue());
-                if (value.compareTo(min) < 0) {
-                    submittedValue = String.valueOf(min.doubleValue());
+        try {
+            if (ComponentUtils.isValueBlank(submittedValue)) {
+                ValueExpression valueExpression = inputNumber.getValueExpression("value");
+                if (valueExpression != null) {
+	                Class<?> type = valueExpression.getType(context.getELContext());
+	                if (type != null && type.isPrimitive() && !ComponentUtils.isValueBlank(inputNumber.getMinValue())) {
+		                // avoid coercion of null or empty string to 0 which may be out of [minValue, maxValue] range
+		                submittedValue = String.valueOf(new BigDecimal(inputNumber.getMinValue()).doubleValue());
+	                }
+	                else {
+		                submittedValue = "";
+	                }
                 }
             }
-            if (!ComponentUtils.isValueBlank(inputNumber.getMaxValue())) {
-                BigDecimal max = new BigDecimal(inputNumber.getMaxValue());
-                if (value.compareTo(max) > 0) {
-                    submittedValue = String.valueOf(max.doubleValue());
+            else {
+                BigDecimal value = new BigDecimal(submittedValue);
+                if (!ComponentUtils.isValueBlank(inputNumber.getMinValue())) {
+                    BigDecimal min = new BigDecimal(inputNumber.getMinValue());
+                    if (value.compareTo(min) < 0) {
+                        submittedValue = String.valueOf(min.doubleValue());
+                    }
+                }
+                if (!ComponentUtils.isValueBlank(inputNumber.getMaxValue())) {
+                    BigDecimal max = new BigDecimal(inputNumber.getMaxValue());
+                    if (value.compareTo(max) > 0) {
+                        submittedValue = String.valueOf(max.doubleValue());
+                    }
                 }
             }
-            
-            inputNumber.setSubmittedValue(submittedValue);
         }
-
+        catch (NumberFormatException ex) { 
```

Comment:
Just by curiosity, after all these input tests. Is `NumberFormatException` is still meant to happen?

Reply:
Yes, NumberFormatException may still happen if client illegally submits non-empty string (e.g. 'crash'). I just thought that wrapping this exception in a FacesException would be more compliant to the PrimeFaces rules, if any.